lib/Url/Profile.php

```diff
@@ -526,7 +526,9 @@ public static function exists($id)
     protected function getDataset($primaryColumnName, $primaryId)
     {
         $query = $this->buildQuery();
-        $query->where($this->getColumnNameWithAlias($primaryColumnName), $primaryId);
+        if($primaryColumnName != '' && $primaryId > 0) {
```

Comment:
$primaryColumnName sollte nie leer sein und die übergeben $primaryId müsste immer größer 0 sein. Ich vermute der Fehler kommt woanders her und wird nur bisher her durchgeschleift.

Reply:
Ja, die Vermutung habe ich auch. Ich habe die vorgelagerten Methoden untersucht. Diese werden nicht nur aus der boot.php aufgerufen, sondern auch wo anders her. Kommt der Aufruf aus der boot.php wird er sauber abgearbeitet, wenn aber wo anders her, kann `$primaryColumnName` leer sein. Ich konnte es nicht anders lösen. Vielleicht hast du eine bessere Idee. In jedem Fall funktioniert es.

Reply:
> Kommt der Aufruf aus der boot.php wird er sauber abgearbeitet, wenn aber wo anders her, kann `$primaryColumnName` leer sein.

Man müsste wissen, wann genau das geschieht.

Reply:
Habs leider nicht mehr auf dem Schirm. Ist schon eine Weile her.

Reply:
Hilft dir der folgende Stacktrace weiter um herauszufinden wann das geschieht? Ich habe den Code gerade aus meiner Testinstallation entfernt und den Stacktrace erzeugen lassen:

**rex_sql_exception:** Error while executing statement "SELECT `data`.`category_id` AS id, `data`.`category_id` AS data_id, `data`.`clang_id` AS data_clang_id, `data`.`seo_description` AS data_seo_description, `data`.`picture` AS data_seo_image, `data`.`updatedate` AS data_sitemap_lastmod, `data`.`seo_title` AS data_seo_title, `data`.`name` AS data_segment_part_1 FROM `rex_d2u_machinery_url_machine_categories` AS data WHERE (data.name != "" AND `data`.`name` IS NOT NULL) AND `data`.`` = :p1 ORDER BY `data_id` asc" using params {"p1":"196"}! SQLSTATE[42S22]: Column not found: 1054 Unknown column 'data.' in 'where clause'
**File:** redaxo/src/core/lib/sql/sql.php
**Line:** 299

<details>
<summary>Stacktrace</summary>

| Function                                     | File                                                                     | Line     |
| -------------------------------------------- | ------------------------------------------------------------------------ | -------- |
| rex_sql->execute                             | redaxo/src/core/lib/sql/sql.php                                          | 341      |
| rex_sql->setQuery                            | redaxo/src/core/lib/sql/sql.php                                          | 1074     |
| rex_sql->getArray                            | redaxo/src/addons/yform/plugins/manager/lib/yform/manager/dataset.php    | 196      |
| rex_yform_manager_dataset::queryCollection   | redaxo/src/addons/yform/plugins/manager/lib/yform/manager/query.php      | 608      |
| rex_yform_manager_query->find                | redaxo/src/addons/url/lib/Url/Profile.php                                | 531      |
| Url\Profile->getDataset                      | redaxo/src/addons/url/lib/Url/Profile.php                                | 248      |
| Url\Profile->buildUrlsByDatasetId            | redaxo/src/addons/url/lib/Url/UrlManager.php                             | 329      |
| Url\UrlManager::getRewriteUrl                | redaxo/src/addons/url/boot.php                                           | 79       |
| rex_package->{closure}                       | redaxo/src/core/lib/extension.php                                        | 45       |
| rex_extension::registerPoint                 | redaxo/src/addons/structure/functions/function_rex_url.php               | 39       |
| rex_getUrl                                   | redaxo/src/addons/d2u_machinery/lib/category.php                         | 580      |
| Category->getURL                             | rex:///module/7/output                                                   | 594      |
| require                                      | redaxo/cache/addons/structure/1.1.content                                | 4        |
| require                                      | redaxo/src/addons/structure/plugins/content/lib/article_content.php      | 102      |
| rex_article_content->getArticle              | rex:///module/61/output                                                  | 5        |
| require                                      | redaxo/cache/addons/structure/1.2.content                                | 4        |
| require                                      | redaxo/src/addons/structure/plugins/content/lib/article_content.php      | 102      |
| rex_article_content->getArticle              | redaxo/cache/addons/templates/4.template                                 | 689      |
| include                                      | rex:///template/3                                                        | 5        |
| require                                      | redaxo/src/addons/structure/plugins/content/lib/article_content_base.php | 421      |
| rex_article_content_base->getArticleTemplate | redaxo/src/addons/structure/plugins/content/boot.php                     | 80       |
| rex_package::{closure}                       | redaxo/src/core/lib/extension.php                                        | 45       |
| rex_extension::registerPoint                 | redaxo/src/core/frontend.php                                             | 26       |
| require                                      | redaxo/src/core/boot.php                                                 | 135      |
| require                                      | index.php                                                                | 9        |

</details>
<details>
<summary>System report (REDAXO 5.8.0, PHP 7.3.6-nmm1)</summary>

| REDAXO        |            |
| ------------: | :--------- |
|       Version | 5.8.0      |


| PHP           |            |
| ------------: | :--------- |
|       Version | 7.3.6-nmm1 |
|       OPcache | no         |
|        Xdebug | no         |


| Database      |                            |
| ------------: | :------------------------- |
|       Version | 5.6.44-nmm1-log            |
| Character set | latin1 (latin1_swedish_ci) |


| Server        |            |
| ------------: | :--------- |
|            OS | Linux      |
|          SAPI | fpm-fcgi   |
|     Webserver | Apache     |


| Request       |              |
| ------------: | :----------- |
|       Browser | Firefox/68.0 |
|      Protocol | HTTP/1.1     |
|         HTTPS | yes          |


| Packages                                         |             |
| -----------------------------------------------: | :---------- |
|                                         be_style | 2.8.0       |
|                              be_style/customizer | 2.8.0       |
|                                  be_style/redaxo | 2.8.0       |
|                                          bloecks | 1.4.2       |
|                                   bloecks/status | 1.4.2       |
|                                   cookie_consent | 3.0.0-beta3 |
|                                          cronjob | 2.5.0       |
|                                      d2u_address | 1.1.1       |
|                                       d2u_helper | 1.5.8       |
|                    d2u_helper/translation_helper | 1.5.8       |
|                                      d2u_history | 1.0.0       |
|                                         d2u_jobs | 1.0.9       |
|                           d2u_jobs/hr4you_import | 1.0.9       |
|                                      d2u_linkbox | 1.2.2       |
|                                    d2u_machinery | 1.3.1       |
|                   d2u_machinery/industry_sectors | 1.3.0       |
|         d2u_machinery/machine_features_extension | 1.3.0       |
| d2u_machinery/machine_steel_processing_extension | 1.3.0       |
|                      d2u_machinery/used_machines | 1.3.1       |
|                                         d2u_news | 1.1.2       |
|                                   d2u_news/fairs | 1.1.2       |
|                                      d2u_partner | 1.0         |
|                                   d2u_references | 1.0.6       |
|                                        d2u_staff | 1.0.2       |
|                                       d2u_videos | 1.0.7       |
|                                  emailobfuscator | 2.1.5       |
|                                          install | 2.5.0       |
|                                    media_manager | 2.7.0       |
|                        media_manager_autorewrite | 1.1.0       |
|                                        mediapool | 2.6.0       |
|                                         metainfo | 2.5.0       |
|                                  multinewsletter | 3.2.5       |
|                                        phpmailer | 2.6.0       |
|                                            sprog | 1.2.0       |
|                                        structure | 2.8.0       |
|                                structure/content | 2.8.0       |
|                                         tinymce4 | 1.2.1       |
|                                              url | 2.0.0-beta3 |
|                                            users | 2.5.2       |
|                                             ycom | 3.0-beta4   |
|                                        ycom/auth | 3.0-beta4   |
|                                       ycom/group | 3.0-beta4   |
|                                  ycom/media_auth | 3.0-beta4   |
|                                            yform | 3.1.1       |
|                                       yform/docs | 3.1.1       |
|                                      yform/email | 3.1.1       |
|                                    yform/manager | 3.1.1       |
|                                         yrewrite | 2.5         |

</details>

Reply:
Nach welcher Spalte lässt du hier filtern? Was ist die `196` ?

Reply:
Die 196 ist der gesuchte Wert in der Spalte `data_id`. Nach welcher Spalte er hier filtern will steht seltsamer Weise nicht da. Vermutlich sollte es die Spalte `data_id` sein.

Reply:
Hast du vielleicht später mal die die Spalten umbenannt? Was steht in deinem Profil links von der 196 ? Es sieht so aus, als ob die Spalte nicht zugewiesen wurde.

Reply:
Du meinst die Datenbankspalten? Die habe ich nicht umbenannt. Links ist die clang_id. Hier ein Screenshot: https://github.com/tbaddade/redaxo_url/pull/137#discussion_r287486743

Reply:
Es geht um diesen Part `AND data.`` = :p1`

Ich müsste ein Filter/Einschränkung sein mit der hinterlegten 196 - `params {"p1":"196"}`

Das müsste so in etwa dann aussehen, wobei bei dir "keine Einschränkung" durch ein Feld ausgewählt ist. Welches wäre das?
<img width="1643" alt="Bildschirmfoto 2019-08-28 um 22 22 50" src="https://user-images.githubusercontent.com/338267/63890020-640ee380-c9e2-11e9-991f-062b4f299812.png">

Reply:
Da keine Einschränkung hinterlegt ist, ist es dann nicht korrekt, dass der Parameter `$primaryColumnName` leer ist? Dann stellt sich eher die Frage warum ein Wert im Parameter `$primaryId` drin steht. Müsste der nicht auch leer sein?

Reply:
Wo kommt denn die 196 her? Siehst du diese in deinem Profil?

Reply:
Die müsste von hier kommen (siehe den Stacktrace): https://github.com/TobiasKrais/d2u_machinery/blob/709496fca9ccf915e6c7f98839e83b8adfd7b53b/lib/category.php#L580

P.S.: bin heute viel unterwegs und werde zum Antworten länger brauchen.

Reply:
Probier mal folgendes: bei mir heißt der Parameter 'category_id'. Häng mal den Parameter an die Artikel URL an und weise ihm einen Wert zu, den es in der Datenbank nicht gibt, z.B. https://test.kaltenbach.com/de/produkte/maschinen/?category_id=196.

Bei mir gibt es zwar diese Kategorie, aber sie ist offline. Deshalb wird aus der Kategorie keine URL erzeugt. Dann tritt der Fehler auf.

Reply:
> Die müsste von hier kommen (siehe den Stacktrace): https://github.com/TobiasKrais/d2u_machinery/blob/709496fca9ccf915e6c7f98839e83b8adfd7b53b/lib/category.php#L580

Erzeugst du so deine Urls für die Instanzen? Diese Variante ist fehleranfällig. Das Generieren der Urls sollte direkt über den Namespace gehen.
`rex_getUrl('', '', ['namespace' => $id])`

Reply:
Ja, genau so mache ich das. Die Variable `$parameterArray` beinhaltet den Namespace als Schlüssel: https://github.com/TobiasKrais/d2u_machinery/blob/709496fca9ccf915e6c7f98839e83b8adfd7b53b/lib/category.php#L578

Oder hab ich deinen Tipp nicht richtig verstanden?

Reply:
Diese Zeile 
https://github.com/TobiasKrais/d2u_machinery/blob/709496fca9ccf915e6c7f98839e83b8adfd7b53b/lib/category.php#L580

sollte so angepasst werden.

`$this->url = rex_getUrl('', '', $parameterArray);`

Dazu sollten diese Schlüssel bei den Profilen hinterlegt sein
- `category_id`
- `used_sale_category_id`
- `used_rent_category_id`


PS: Diese Zeile kann wegfallen https://github.com/TobiasKrais/d2u_machinery/blob/709496fca9ccf915e6c7f98839e83b8adfd7b53b/lib/category.php#L566-L568 oder dein `else` hat einen falschen Key

Reply:
Danke! Die Schlüssel sind alle in Profilen hinterlegt. Und du hast Recht, die Zeilen ist absolut unnötig. Danke für den Hinweis.

Die `article_id` und `clang_id` übergebe ich deshalb, weil es sein kann, dass ich mit einem Objekt in einer Sprache arbeite, die noch nicht online ist. Dann werden auch keine URLs vom URL Addon generiert (so habe ich es in dem zugrunde liegenden VIEW per JOIN festgelegt). Es muss dann eine URL wie schon oben genannt (z.B. https://test.kaltenbach.com/de/produkte/maschinen/?category_id=196) generiert werden. In den Einstellungen des Addons gibt es entsprechende Stellschrauben.

Zurück zum eigentlichen Problem: der Fehler des Issues tritt dann auf, wenn:
- der Schlüssel in einem Profil hinterlegt ist und
- die ID aber nicht in der URL Tabelle vorhanden ist.
Wenn ich dich richtig verstanden habe, ist der `$primaryColumnName` an der Stelle im Code korrekter weise leer, aber die `$primaryId` sollte eben auch leer oder 0 sein, ist sie aber nicht. Richtig? Wenn ja, müsste der Fehler und die Fehlerbehebung vorgelagert im Stacktrace erfolgen.

Reply:
Ziel der Version 2 soll sein, dass man die Urls nur über den Namespace/Schlüssel erzeugt und nicht über die Artikel-Id und Clang-Id. Das war in der 1er Version sehr fehleranfällig.

Da du scheinbar keinen Schlüssel hier übergibst sondern nur die Datensatz-Id kommt der Stacktrace. Ich schau mir das mal an und lasse dann eine Exception ausgeben.

> Die article_id und clang_id übergebe ich deshalb, weil es sein kann, dass ich mit einem Objekt in einer Sprache arbeite, die noch nicht online ist. Dann werden auch keine URLs vom URL Addon generiert

Das sollte natürlich nicht sein. Die Urls sollten für online und offline Sprachen erstellt werden.

Reply:
> Ich schau mir das mal an und lasse dann eine Exception ausgeben.

Das geht gar nicht :) … Ich kann die ParamKeys nur mit den Schlüsseln die beim Profil hinterlegt sind vergleichen und wenn dann ein Treffer vorhanden ist, die Url zurückgeben. Alle anderen ParamKeys können ja gewollte sein.

`rex_getUrl('', '', ['profile-key' => 196, 'category_id' => 196])`;

Wenn von diesen 2 Paramkeys keiner in den Profilen hinterlegt wurde, wird die Url von yrewrite zurückgegeben.

Und die kann so sein
`/start/de/?profile-key=196&category_id=196`
oder so, da die Artikel-Id fehlt
`/produkte/de/?profile-key=196&category_id=196`

Du verstehst? Deshalb ist es wichtig den Namespace/Schlüssel zu hinterlegen und auch beim Aufruf von rex_getUrl zu hinterlegen.

Reply:
> > Die article_id und clang_id übergebe ich deshalb, weil es sein kann, dass ich mit einem Objekt in einer Sprache arbeite, die noch nicht online ist. Dann werden auch keine URLs vom URL Addon generiert
> 
> Das sollte natürlich nicht sein. Die Urls sollten für online und offline Sprachen erstellt werden.

Tauchen die URLs dann nicht auch in der sitemap.xml auf? Das wollte ich damit verhindern.

Reply:
Das erstellen der Urls sollte unabhängig vom Status der Sprache sein. Das Ausliefern an die Sitemap müsste dann nochmal geprüft werden.

Reply:
> Da du scheinbar keinen Schlüssel hier übergibst sondern nur die Datensatz-Id kommt der Stacktrace.

Der Schlüssel heißt 'category_id' und wird übergeben. Nur er "verschwindet" unterwegs, und zwar genau hier: https://github.com/tbaddade/redaxo_url/blob/ae75a4fb56e03c4cdecfc5f00bd24d52127b6026/lib/Url/UrlManager.php#L341, noch genauer an der Stelle des Codes: `\rex_sql_table::get($profile->getTableName())->getPrimaryKey()[0]`. Ein paar Zeilen weiter oben (https://github.com/tbaddade/redaxo_url/blob/ae75a4fb56e03c4cdecfc5f00bd24d52127b6026/lib/Url/UrlManager.php#L334) ist der Wert der Variable `$urlParamKey` noch 'category_id'.

Reply:
Hmm, `$profile->getTableName())->getPrimaryKey()[0]` kann auch gar keinen Wert liefern. Die Zugrunde liegende Tabelle ist keine Tabelle, sondern ein VIEW und hat keinen Primary Key...

Reply:
Funktioniert das ganze ohne View?

Reply:
Leider Nein. Noch schlimmer: mein Patch ist leider wirkungslos. Warum? `$profile->getTableName())->getPrimaryKey()[0]` ermittelt nun den korrekten Primärschlüssel. Damit haben in der Methode `Profile::getDataset()` beide Parameter einen Wert :-( und mein Patch prüft, ob einer der Parameter nicht korrekt ist.

Reply:
Wozu braucht es die Zeile https://github.com/tbaddade/redaxo_url/blob/ae75a4fb56e03c4cdecfc5f00bd24d52127b6026/lib/Url/Profile.php#L529 eigentlich? Wenn ich sie entferne ist der Fehler weg und alles funktioniert bei mir wie gewünscht.

Reply:
> Wozu braucht es die Zeile

um den Datensatz zu finden.

> Damit haben in der Methode Profile::getDataset() beide Parameter einen Wert

So sollte es ja auch sein. Jetzt muss man schauen warum die Urls nicht gefunden werden.
Bekommst du einen Whoops? Liefert dir rex_sql_exception etwas?

Reply:
> um den Datensatz zu finden.

Es funktioniert auch ohne :-)

Reply:
> Bekommst du einen Whoops? Liefert dir rex_sql_exception etwas?

Genau das gleiche wie oben:
rex_sql_exception thrown with message "Error while executing statement "SELECT `data`.`country_id` AS id, `data`.`country_id` AS data_id, `data`.`clang_id` AS data_clang_id, `data`.`name` AS data_segment_part_1, `data`.`country_id` AS data_segment_part_2 FROM `rex_d2u_address_countries_lang` AS data WHERE (data.name != "" AND `data`.`name` IS NOT NULL OR data.country_id != "" AND `data`.`country_id` IS NOT NULL) AND `data`.`` = :p1 ORDER BY `data_id` asc" using params {"p1":10000}! SQLSTATE[42S22]: Column not found: 1054 Unknown column 'data.' in 'where clause'"

Stacktrace:
#20 rex_sql_exception in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/core/lib/sql/sql.php:299
#19 PDOException in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/core/lib/sql/sql.php:296
#18 PDOStatement:execute in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/core/lib/sql/sql.php:296
#17 rex_sql:execute in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/core/lib/sql/sql.php:341
#16 rex_sql:setQuery in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/core/lib/sql/sql.php:1074
#15 rex_sql:getArray in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/addons/yform/plugins/manager/lib/yform/manager/dataset.php:196
#14 rex_yform_manager_dataset:queryCollection in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/addons/yform/plugins/manager/lib/yform/manager/query.php:608
#13 rex_yform_manager_query:find in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/addons/url/lib/Url/Profile.php:531
#12 Url\Profile:getDataset in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/addons/url/lib/Url/Profile.php:248
#11 Url\Profile:buildUrlsByDatasetId in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/addons/url/lib/Url/UrlManager.php:330
#10 Url\UrlManager:getRewriteUrl in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/addons/url/boot.php:80
#9 rex_package:{closure} in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/core/lib/extension.php:45
#8 rex_extension:registerPoint in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/addons/structure/functions/function_rex_url.php:41
#7 rex_getUrl in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/cache/addons/templates/4.template:2
#6 include in rex:///template/2:5
#5 require in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/addons/structure/plugins/content/lib/article_content_base.php:421
#4 rex_article_content_base:getArticleTemplate in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/addons/structure/plugins/content/boot.php:80
#3 rex_package:{closure} in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/core/lib/extension.php:45
#2 rex_extension:registerPoint in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/core/frontend.php:26
#1 require in /www/htdocs/w01278ee/test.kaltenbach.com/redaxo/src/core/boot.php:135
#0 require in /www/htdocs/w01278ee/test.kaltenbach.com/index.php:9